<a href="https://colab.research.google.com/github/kittimaxz/DWDM21/blob/main/ProjectFinal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **สมาชิก**
**Sandboxนะจ๊ะ**
1. 623020516-1 นางสาวชนัญชิดา เมธีกุลมานิต
2. 623020521-8 นางสาวณิชากร ไชยสุวรรณ
3. 623021040-9 นางสาวกิตติมา อุปสุข
4. 623021050-6 นางสาวนฤมล ไสยโสภณ
5. 623021052-2 นางสาวภิญญาดา เพ็ญสุข 

###แหล่งที่มาของข้อมูล
* ย่านชุมชนเก่า แหล่งที่มาของข้อมูล https://data.go.th/dataset/old-community?fbclid=IwAR0TJxEAwTtBKYS853iy-LVEDE09z9WawwXii0BufdHWwojpU0_Lvw3ciYM
* ผู้เยี่ยมเยือน จำแนกตามจังหวัด แหล่งท่มาของข้อมูล https://data.go.th/dataset/visitors?fbclid=IwAR0mLjK4IUjLWa99UBhG5sM6FGGnicg_iVRzQHTp8btujsUbFEOzmGFaRYA
* รายได้จากการท่องเที่ยว จำแนกรายจังหวัด https://data.go.th/dataset/1559?fbclid=IwAR3-ZnxUzDyap-vs3T3vdkkwXgIu5ye3pwha8okKoZk6MQMhYl6KNmYeu9M

In [1]:
import pandas as pd # ทำงานกับข้อมูลลักษณะตาราง

In [2]:
from google.colab import drive # package เชื่อม google  drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
ls

drive/  sample_data/


In [4]:
import os # จัดการไฟล์และ path (ที่อยู่ของไฟล์)

In [5]:
path = '/content/drive/My Drive/DWDM2021_DATA'

### นำเข้าข้อมูล

#### ข้อมูลที่ 1 ย่านชุมชนเก่า

In [6]:
data_1 = pd.read_csv(os.path.join(path,'old-community-area-updated-may-2021.csv')) # ย่านชุมชนเก่า
data_1 

ลำดับที่  ...                                        ลักษณะสำคัญ
0           1  ...  นครชุม เป็นชุมชนตลาดการค้า ที่มีประวัติความเป็...
1           2  ...  บ้านพรานกระต่าย เป็นชุมชนเกษตรกรรมชนบท ที่มีเร...
2           3  ...  บ้านห้วยแม่ซ้าย เป็นชุมชนเกษตรกรรม และชุมชนชาต...
3           4  ...  บ้านถ้ำผาตอง เป็นชุมชนเกษตรกรรม และชุมชนกลุ่มอ...
4           5  ...  บ้านกะเหรี่ยงรวมมิตร เป็นชุมชนเกษตรกรรม ชุมชนต...
..        ...  ...                                                ...
608       609  ...  ชุมชนบ้านปะอาว เป็นชุมชนชนบทตั้งอยู่บนที่ดอน พ...
609       610  ...  ชุมชนพาณิชยกรรมเก่า เป็นย่านพาณิชยกรรมเก่าแห่ง...
610       611  ...  ชุมชนแก่งสะพือ เป็นย่านชุมชนเก่าริมน้ำแม่น้ำมู...
611       612  ...  ชุมชนบ้านแคน เป็นชุมชนชนบทตั้งอยู่บนที่ดอน พื้...
612       613  ...  ชุมชนบ้านชาด เป็นชุมชนเก่าตั้งอยู่ใกล้กับแหล่ง...

[613 rows x 9 columns]

**ตาราง data_city คือ ข้อมูลย่านชุมชนเก่า**

**คำอธิบายตาราง**

*   ชื่อย่านชุมชนเก่า คือ ชื่อของชุมชนเก่า
*   ตำบล คือ ที่ตั้งของชุมชนเก่า
* อำเภอ คือ ที่ตั้งของชุมชนเก่า
* จังหวัด คือ เขตที่ตั้งของชุมชนเก่า
* ภาค คือ ภูมิประเทศที่ตั้งของชุมชนเก่า
* East คือ ลองจิจูดองศาตะวันออกที่ตั้งของชุมชนเก่า
* North คือ ละติจูดองศาเหนือที่ตั้งของชุมชนเก่า
* ลักษณะสำคัญ คือ รายละเอียดของชุมชนเก่า

In [7]:
data_city = data_1[['ชื่อย่านชุมชนเก่า','ตำบล','อำเภอ','จังหวัด','ภาค','East','North','ลักษณะสำคัญ']] # ย่านชุมชนเก่า
data_city

ชื่อย่านชุมชนเก่า  ...                                        ลักษณะสำคัญ
0                        นครชุม  ...  นครชุม เป็นชุมชนตลาดการค้า ที่มีประวัติความเป็...
1               บ้านพรานกระต่าย  ...  บ้านพรานกระต่าย เป็นชุมชนเกษตรกรรมชนบท ที่มีเร...
2    บ้านห้วยแม่ซ้าย (บ้านยะฟู)  ...  บ้านห้วยแม่ซ้าย เป็นชุมชนเกษตรกรรม และชุมชนชาต...
3                  บ้านถ้ำผาตอง  ...  บ้านถ้ำผาตอง เป็นชุมชนเกษตรกรรม และชุมชนกลุ่มอ...
4          บ้านกะเหรี่ยงรวมมิตร  ...  บ้านกะเหรี่ยงรวมมิตร เป็นชุมชนเกษตรกรรม ชุมชนต...
..                          ...  ...                                                ...
608              ชุมชนบ้านปะอาว  ...  ชุมชนบ้านปะอาว เป็นชุมชนชนบทตั้งอยู่บนที่ดอน พ...
609         ชุมชนพาณิชยกรรมเก่า  ...  ชุมชนพาณิชยกรรมเก่า เป็นย่านพาณิชยกรรมเก่าแห่ง...
610              ชุมชนแก่งสะพือ  ...  ชุมชนแก่งสะพือ เป็นย่านชุมชนเก่าริมน้ำแม่น้ำมู...
611                ชุมชนบ้านแคน  ...  ชุมชนบ้านแคน เป็นชุมชนชนบทตั้งอยู่บนที่ดอน พื้...
612                ชุมชนบ้านชาด  ...  ชุมชนบ้านชาด เป็นชุมชนเก่าตั้งอยู่ใกล้กับแหล่ง...

[613 rows x 8 columns]

ตรวจสอบค่า missing ของข้อมูลตาราง data_city

In [8]:
data_city.isnull().any() 

ชื่อย่านชุมชนเก่า    False
ตำบล                 False
อำเภอ                False
จังหวัด              False
ภาค                  False
East                 False
North                False
ลักษณะสำคัญ          False
dtype: bool

แสดงผลเป็น False ทั้งหมด แปลว่า ไม่มีค่า missing ในข้อมูล ย่านชุมชนเก่า

#### ข้อมูลที่ 2 ผู้เยี่ยมเยือน จำแนกตามจังหวัด

In [9]:
data_2= pd.read_csv(os.path.join(path,'visitors.csv'))  # ผู้เยี่ยมเยือน จำแนกตามจังหวัด
data_2

จังหวัด รหัสจังหวัด        2549  ...      2560      2561  Unnamed: 15
0     กรุงเทพมหานคร       th-bm  36172138.0  ...  63575737  65534281          NaN
1       สมุทรปราการ       th-sp    958352.0  ...   3218934   3373639          NaN
2           นนทบุรี       th-no    830593.0  ...   2514677   2664025          NaN
3          ปทุมธานี       th-pt    868449.0  ...   2014281   2098344          NaN
4   พระนครศรีอยุธยา       th-pa   3373929.0  ...   7631557   8349613          NaN
..              ...         ...         ...  ...       ...       ...          ...
72             ตรัง       th-tg    726373.0  ...   1514092   1578906          NaN
73           พัทลุง       th-pl    448000.0  ...   1569333   1641841          NaN
74          ปัตตานี       th-pi    160320.0  ...    272739    281988          NaN
75             ยะลา       th-yl    282463.0  ...    657317    680018          NaN
76         นราธิวาส       th-nw    336473.0  ...    665710    692057          NaN

[77 rows x 16 columns]

ตัดตารางข้อมูลของ data_2 เลือกเฉพาะข้อมูลปี 2555-2561

In [10]:
data_visitor = data_2[['จังหวัด','รหัสจังหวัด','2555','2556','2557','2558','2559','2560','2561']] # ผู้เยี่ยมเยือน จำแนกตามจังหวัด
data_visitor

จังหวัด รหัสจังหวัด      2555  ...      2559      2560      2561
0     กรุงเทพมหานคร       th-bm  47185031  ...  59196331  63575737  65534281
1       สมุทรปราการ       th-sp   2107433  ...   2900180   3218934   3373639
2           นนทบุรี       th-no   1663990  ...   2090095   2514677   2664025
3          ปทุมธานี       th-pt   1474292  ...   1896549   2014281   2098344
4   พระนครศรีอยุธยา       th-pa   6066545  ...   7216514   7631557   8349613
..              ...         ...       ...  ...       ...       ...       ...
72             ตรัง       th-tg   1137961  ...   1465811   1514092   1578906
73           พัทลุง       th-pl   1089992  ...   1487188   1569333   1641841
74          ปัตตานี       th-pi    214285  ...    257143    272739    281988
75             ยะลา       th-yl    429044  ...    624879    657317    680018
76         นราธิวาส       th-nw    489944  ...    646020    665710    692057

[77 rows x 9 columns]

เปลี่ยนชื่อ column (data_visitor)

In [11]:
data_visitor.rename(columns={'2555':'2555visitor'},inplace=True)
data_visitor.rename(columns={'2556':'2556visitor'},inplace=True)
data_visitor.rename(columns={'2557':'2557visitor'},inplace=True)
data_visitor.rename(columns={'2558':'2558visitor'},inplace=True)
data_visitor.rename(columns={'2559':'2559visitor'},inplace=True)
data_visitor.rename(columns={'2560':'2560visitor'},inplace=True)
data_visitor.rename(columns={'2561':'2561visitor'},inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


**ตาราง data_visitor คือ ข้อมูลผู้เยี่ยมเยือน จำแนกตามจังหวัด**

**คำอธิบายตาราง**

* จังหวัด คือ ที่ตั้งของแหล่งท่องเที่ยว
* รหัสจังหวัด คือ รหัสประจำของจังหวัดนั้นๆ 
* 2555visitor คือ จำนวนผู้ที่มาเยี่ยมชมแหล่งท่องเที่ยวในจังหวัดนั้นๆ 
* 2556visitor คือ จำนวนผู้ที่มาเยี่ยมชมแหล่งท่องเที่ยวในจังหวัดนั้นๆ 
* 2557visitor คือ จำนวนผู้ที่มาเยี่ยมชมแหล่งท่องเที่ยวในจังหวัดนั้นๆ 
* 2558visitor คือ จำนวนผู้ที่มาเยี่ยมชมแหล่งท่องเที่ยวในจังหวัดนั้นๆ 
* 2559visitor คือ จำนวนผู้ที่มาเยี่ยมชมแหล่งท่องเที่ยวในจังหวัดนั้นๆ 
* 2560visitor คือ จำนวนผู้ที่มาเยี่ยมชมแหล่งท่องเที่ยวในจังหวัดนั้นๆ 
* 2561visitor คือ จำนวนผู้ที่มาเยี่ยมชมแหล่งท่องเที่ยวในจังหวัดนั้นๆ 

In [12]:
data_visitor

จังหวัด รหัสจังหวัด  ...  2560visitor  2561visitor
0     กรุงเทพมหานคร       th-bm  ...     63575737     65534281
1       สมุทรปราการ       th-sp  ...      3218934      3373639
2           นนทบุรี       th-no  ...      2514677      2664025
3          ปทุมธานี       th-pt  ...      2014281      2098344
4   พระนครศรีอยุธยา       th-pa  ...      7631557      8349613
..              ...         ...  ...          ...          ...
72             ตรัง       th-tg  ...      1514092      1578906
73           พัทลุง       th-pl  ...      1569333      1641841
74          ปัตตานี       th-pi  ...       272739       281988
75             ยะลา       th-yl  ...       657317       680018
76         นราธิวาส       th-nw  ...       665710       692057

[77 rows x 9 columns]

ตรวจสอบค่า missing ของข้อมูลตาราง data_visitor

In [13]:
data_visitor.isnull().any()

จังหวัด        False
รหัสจังหวัด    False
2555visitor    False
2556visitor    False
2557visitor    False
2558visitor    False
2559visitor    False
2560visitor    False
2561visitor    False
dtype: bool

แสดงผลเป็น false ทั้งหมด หมายความว่า ไม่มีค่า Missing ในข้อมูล ผู้เยี่ยมเยือน จำแนกตามจังหวัด

#### ข้อมููลที่ 3 รายได้จากการท่องเที่ยว จำแนกตามจังหวัด

In [14]:
data_3= pd.read_csv(os.path.join(path,'revenue.csv')) # รายได้จากการท่องเที่ยว จำแนกตามจังหวัด
data_3

จังหวัด รหัสจังหวัด       2555  ...       2559          2560        2561
0   กรุงเทพมหานคร       th-bm  570327.58  ...  822454.21  947946.30000  1051118.51
1       กาญจนบุรี       th-kn   11910.00  ...   20404.73   24440.77288    27138.35
2        จันทบุรี       th-ct    4214.00  ...    5630.03    7628.47000     8469.45
3      ฉะเชิงเทรา       th-cc    3571.69  ...    4240.41    4655.55000     5007.28
4          ชลบุรี       th-cb  100537.00  ...  187788.42  240610.45000   272435.77
..            ...         ...        ...  ...        ...           ...         ...
72          ระนอง       th-rn    2780.00  ...    3954.75    4311.07000     4706.13
73          สงขลา       th-sg   26702.00  ...   53493.02   59831.77000    68097.42
74           สตูล       th-sa    5332.00  ...    7438.85    8168.70000     9101.47
75   สุราษฎร์ธานี       th-st   35312.00  ...   76546.39   92228.01000   104503.71
76         บึงกาฬ       th-bk       0.00  ...     900.50     976.61000     1047.13

[77 rows x 9 columns]

ตัดตารางข้อมูลของ data_3 เลือกเฉพาะข้อมูลปี 2555-2561

In [15]:
data_income = data_3[['รหัสจังหวัด','2555','2556','2557','2558','2559','2560','2561']] #รายได้จากการท่องเที่ยว จำแนกตามจังหวัด
data_income

รหัสจังหวัด       2555       2556  ...       2559          2560        2561
0        th-bm  570327.58  626535.05  ...  822454.21  947946.30000  1051118.51
1        th-kn   11910.00   13417.00  ...   20404.73   24440.77288    27138.35
2        th-ct    4214.00    4569.00  ...    5630.03    7628.47000     8469.45
3        th-cc    3571.69    3735.06  ...    4240.41    4655.55000     5007.28
4        th-cb  100537.00  111109.00  ...  187788.42  240610.45000   272435.77
..         ...        ...        ...  ...        ...           ...         ...
72       th-rn    2780.00    3250.00  ...    3954.75    4311.07000     4706.13
73       th-sg   26702.00   37276.00  ...   53493.02   59831.77000    68097.42
74       th-sa    5332.00    5905.00  ...    7438.85    8168.70000     9101.47
75       th-st   35312.00   51550.00  ...   76546.39   92228.01000   104503.71
76       th-bk       0.00       0.00  ...     900.50     976.61000     1047.13

[77 rows x 8 columns]

เปลี่ยนชื่อ column (data_income)

In [16]:
data_income.rename(columns={'2555':'2555income'},inplace=True)
data_income.rename(columns={'2556':'2556income'},inplace=True)
data_income.rename(columns={'2557':'2557income'},inplace=True)
data_income.rename(columns={'2558':'2558income'},inplace=True)
data_income.rename(columns={'2559':'2559income'},inplace=True)
data_income.rename(columns={'2560':'2560income'},inplace=True)
data_income.rename(columns={'2561':'2561income'},inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


**ตาราง data_income คือ รายได้จากการท่องเที่ยว จำแนกตามจังหวัด**

**คำอธิบายตาราง**

* รหัสจังหวัด คือ รหัสประจำของจังหวัดนั้นๆ 
* 2555income คือ รายได้จากผูู้ที่มาเยี่ยมชมแหล่งท่องเที่ยวในจังหวัดนั้นๆ 
* 2556income คือ รายได้จากผูู้ที่มาเยี่ยมชมแหล่งท่องเที่ยวในจังหวัดนั้นๆ  
* 2557income คือ รายได้จากผูู้ที่มาเยี่ยมชมแหล่งท่องเที่ยวในจังหวัดนั้นๆ 
* 2558income คือ รายได้จากผูู้ที่มาเยี่ยมชมแหล่งท่องเที่ยวในจังหวัดนั้นๆ 
* 2559income คือ รายได้จากผูู้ที่มาเยี่ยมชมแหล่งท่องเที่ยวในจังหวัดนั้นๆ 
* 2560income คือ รายได้จากผูู้ที่มาเยี่ยมชมแหล่งท่องเที่ยวในจังหวัดนั้นๆ 
* 2561income คือ รายได้จากผูู้ที่มาเยี่ยมชมแหล่งท่องเที่ยวในจังหวัดนั้นๆ 

In [17]:
data_income

รหัสจังหวัด  2555income  2556income  ...  2559income    2560income  2561income
0        th-bm   570327.58   626535.05  ...   822454.21  947946.30000  1051118.51
1        th-kn    11910.00    13417.00  ...    20404.73   24440.77288    27138.35
2        th-ct     4214.00     4569.00  ...     5630.03    7628.47000     8469.45
3        th-cc     3571.69     3735.06  ...     4240.41    4655.55000     5007.28
4        th-cb   100537.00   111109.00  ...   187788.42  240610.45000   272435.77
..         ...         ...         ...  ...         ...           ...         ...
72       th-rn     2780.00     3250.00  ...     3954.75    4311.07000     4706.13
73       th-sg    26702.00    37276.00  ...    53493.02   59831.77000    68097.42
74       th-sa     5332.00     5905.00  ...     7438.85    8168.70000     9101.47
75       th-st    35312.00    51550.00  ...    76546.39   92228.01000   104503.71
76       th-bk        0.00        0.00  ...      900.50     976.61000     1047.13

[77 rows x 8 columns]

ตรวจสอบค่า missing ของข้อมูลตาราง data_income

In [18]:
data_income.isnull().any()

รหัสจังหวัด    False
2555income     False
2556income     False
2557income     False
2558income     False
2559income     False
2560income     False
2561income     False
dtype: bool

แสดงผลเป็น false ทั้งหมด หมายความว่า ไม่มีค่า missing ในข้อมูล รายได้จากการท่องเที่ยว จำแนกตามจังหวัด

### รวม 2 ตารางเข้าด้วยกัน .merge()

เชื่อมตาราง data_visitor กับตาราง data_income เข้าด้วยกันด้วย 'รหัสจังหวัด'

In [19]:
merged_data_visitor_data_income = data_visitor.merge(data_income,how='left',left_on='รหัสจังหวัด',right_on='รหัสจังหวัด') 
merged_data_visitor_data_income

จังหวัด รหัสจังหวัด  ...  2560income  2561income
0     กรุงเทพมหานคร       th-bm  ...   947946.30  1051118.51
1       สมุทรปราการ       th-sp  ...     6144.65     6656.34
2           นนทบุรี       th-no  ...     3963.09     4368.99
3          ปทุมธานี       th-pt  ...     3172.61     3443.06
4   พระนครศรีอยุธยา       th-pa  ...    16901.48    18718.90
..              ...         ...  ...         ...         ...
72             ตรัง       th-tg  ...     9028.01     9916.29
73           พัทลุง       th-pl  ...     3155.92     3489.08
74          ปัตตานี       th-pi  ...      996.16     1100.97
75             ยะลา       th-yl  ...     3145.57     3443.17
76         นราธิวาส       th-nw  ...     2888.53     3137.67

[77 rows x 16 columns]

In [20]:
merged_data_visitor_data_income.isnull().any() # ตรวจสอบค่า missing

จังหวัด        False
รหัสจังหวัด    False
2555visitor    False
2556visitor    False
2557visitor    False
2558visitor    False
2559visitor    False
2560visitor    False
2561visitor    False
2555income     False
2556income     False
2557income     False
2558income     False
2559income     False
2560income     False
2561income     False
dtype: bool

แสดงผลเป็น false ทั้งหมด หมายความว่า ไม่มีค่า missing ในข้อมูล จากการรวมตารางของ รายได้ และ ผู้เยี่ยมเยือนเข้าด้วยกัน

#### หลังจากผ่านการตรวจสอบ missing แล้ว นำข้อมูลที่ได้จากการรวมตารางครั้งแรกมาเชื่อมกับตาราง data_city 

In [21]:
data_mix = merged_data_visitor_data_income.merge(data_city,how='left',left_on='จังหวัด',right_on='จังหวัด')
data_mix

จังหวัด  ...                                        ลักษณะสำคัญ
0    กรุงเทพมหานคร  ...                                                NaN
1      สมุทรปราการ  ...  ชุมชนเมืองนครเขื่อนขันธ์ เป็นชุมชนย่านการค้าเก...
2      สมุทรปราการ  ...  ชุมชนตลาดบางพลี เป็นชุมชนย่านตลาดการค้าเก่าแก่...
3      สมุทรปราการ  ...  ชุมชนตลาดคลองสวน เป็นชุมชนย่านตลาดการค้า บริเว...
4      สมุทรปราการ  ...  ชุมชนบ้านสาขลา เป็นชุมชนประมงของชาวนาเกลือ ที่...
..             ...  ...                                                ...
609       นราธิวาส  ...  ดุซงญอ เป็นชุมชนตลาด และชุมชนความทรงจำ "ดุซงญอ...
610       นราธิวาส  ...  ตะโละมาเนาะ เป็นชุมชนเกษตรกรรม และชุมชนโดยรอบโ...
611       นราธิวาส  ...  ละหาร เป็นชุมชนเกษตรกรรม บ้านละหารตั้งอยู่ในอำ...
612       นราธิวาส  ...  ตลาดยี่งอ เป็นชุมชนตลาด "เมืองยี่งอ" อาจจะเรีย...
613       นราธิวาส  ...  ตลาดสุไหงปาดี เป็นชุมชนตลาด และชุมชนโดยรอบทางร...

[614 rows x 23 columns]

In [22]:
data_mix.isnull().any() # ตรวจสอบค่า missing

จังหวัด              False
รหัสจังหวัด          False
2555visitor          False
2556visitor          False
2557visitor          False
2558visitor          False
2559visitor          False
2560visitor          False
2561visitor          False
2555income           False
2556income           False
2557income           False
2558income           False
2559income           False
2560income           False
2561income           False
ชื่อย่านชุมชนเก่า     True
ตำบล                  True
อำเภอ                 True
ภาค                   True
East                  True
North                 True
ลักษณะสำคัญ           True
dtype: bool

ในการแสดงผล พบว่ามีข้อมูล ชื่อย่านชุมชนเก่า,ตำบล,อำเภอ,ภาค,East,North ,ลักษณะสำคัญ ที่ยังมีค่า missing อยู่ จึงต้องทำการจัดการค่า Missing ก่อน

#### จัดการค่า missing ด้วยวิธี Handling missing (แทนด้วย calss ใหม่ (unknow) ) โดยใช้คำสั่ง .fillna()

In [23]:
data_mix[['ชื่อย่านชุมชนเก่า']] = data_mix[['ชื่อย่านชุมชนเก่า']].fillna('ไม่มีข้อมูล')
data_mix[['ตำบล']] =data_mix[['ตำบล']].fillna('ไม่มีข้อมูล')
data_mix[['อำเภอ']] = data_mix[['อำเภอ']].fillna('ไม่มีข้อมูล')
data_mix[['ภาค']] = data_mix[['ภาค']].fillna('ไม่มีข้อมูล')
data_mix[['East']] =data_mix[['East']].fillna('ไม่มีข้อมูล')
data_mix[['North']] =data_mix[['North']].fillna('ไม่มีข้อมูล')
data_mix[['ลักษณะสำคัญ']] =data_mix[['ลักษณะสำคัญ']].fillna('ไม่มีข้อมูล')

In [24]:
data_mix.isnull().any() # ตรวจสอบอีกรอบ

จังหวัด              False
รหัสจังหวัด          False
2555visitor          False
2556visitor          False
2557visitor          False
2558visitor          False
2559visitor          False
2560visitor          False
2561visitor          False
2555income           False
2556income           False
2557income           False
2558income           False
2559income           False
2560income           False
2561income           False
ชื่อย่านชุมชนเก่า    False
ตำบล                 False
อำเภอ                False
ภาค                  False
East                 False
North                False
ลักษณะสำคัญ          False
dtype: bool

In [25]:
data_mix

จังหวัด  ...                                        ลักษณะสำคัญ
0    กรุงเทพมหานคร  ...                                        ไม่มีข้อมูล
1      สมุทรปราการ  ...  ชุมชนเมืองนครเขื่อนขันธ์ เป็นชุมชนย่านการค้าเก...
2      สมุทรปราการ  ...  ชุมชนตลาดบางพลี เป็นชุมชนย่านตลาดการค้าเก่าแก่...
3      สมุทรปราการ  ...  ชุมชนตลาดคลองสวน เป็นชุมชนย่านตลาดการค้า บริเว...
4      สมุทรปราการ  ...  ชุมชนบ้านสาขลา เป็นชุมชนประมงของชาวนาเกลือ ที่...
..             ...  ...                                                ...
609       นราธิวาส  ...  ดุซงญอ เป็นชุมชนตลาด และชุมชนความทรงจำ "ดุซงญอ...
610       นราธิวาส  ...  ตะโละมาเนาะ เป็นชุมชนเกษตรกรรม และชุมชนโดยรอบโ...
611       นราธิวาส  ...  ละหาร เป็นชุมชนเกษตรกรรม บ้านละหารตั้งอยู่ในอำ...
612       นราธิวาส  ...  ตลาดยี่งอ เป็นชุมชนตลาด "เมืองยี่งอ" อาจจะเรีย...
613       นราธิวาส  ...  ตลาดสุไหงปาดี เป็นชุมชนตลาด และชุมชนโดยรอบทางร...

[614 rows x 23 columns]

**จังหวัดกาญจนบุรี มีผู้มาเยี่ยมชมย่านชุมชนเก่าในปี 2560 กี่คน 2560visitor**

**จังหวัดกาญจนบุรี มีรายได้จากการมาเยี่ยมชมชุมชนเก่าในปี 2560 กี่บาท 2560income**

**จังหวัดเชียวใหม่มีย่านชุมชนเก่าใดบ้าง**

In [26]:
Ex1 =data_mix[data_mix['ชื่อย่านชุมชนเก่า']== 'นครชุม']  
Ex1[['จังหวัด','ชื่อย่านชุมชนเก่า','2560visitor','2560income']]

จังหวัด ชื่อย่านชุมชนเก่า  2560visitor  2560income
320  กำแพงเพชร            นครชุม       752553      1529.1

In [27]:
Ex2 = data_mix[data_mix['จังหวัด']== 'เชียงใหม่']  
Ex2[['จังหวัด','ชื่อย่านชุมชนเก่า']] #จังหวัดเชียงใหม่ มีย่านชุมชนเก่าใดบ้าง

จังหวัด                         ชื่อย่านชุมชนเก่า
196  เชียงใหม่                       ย่านการค้า ถนนท่าแพ
197  เชียงใหม่                   ย่านการค้า ถนนวิชยานนท์
198  เชียงใหม่  ย่านการค้าถนนเจริญเมือง และย่านสถานีรถไฟ
199  เชียงใหม่              ย่านถนนราชมรรคาและถนนสิงหราช
200  เชียงใหม่                   ย่านสี่แยกตลาดช้างเผือก
201  เชียงใหม่                              ย่านวัดเกตุ 
202  เชียงใหม่                                บ้านดอยปุย
203  เชียงใหม่                              ทรายมูลเมือง
204  เชียงใหม่                                บ้านวัวลาย
205  เชียงใหม่                                   บ้านฮ่อ
206  เชียงใหม่                                   พวกแต้ม
207  เชียงใหม่                            บ้านช่างเคี่ยน
208  เชียงใหม่            ย่านการค้าสี่แยกอำเภอดอยสะเก็ด
209  เชียงใหม่                                 บ้านปางบง
210  เชียงใหม่                               ตลาดสันทราย
211  เชียงใหม่                             บ้านสันป่าสัก
212  เชียงใหม่                              ตลาดสันกำแพง
213  เชียงใหม่                                 บ้านปงไหว
214  เชียงใหม่                             ตลาดสันป่าตอง
215  เชียงใหม่                            บ้านทุ่งเสี้ยว
216  เชียงใหม่                            บ้านต้นแหนน้อย
217  เชียงใหม่                     ตลาดรอบสถานีรถไฟสารภี
218  เชียงใหม่                               เวียงกุมกาม
219  เชียงใหม่                             บ้านวัดจันทร์
220  เชียงใหม่                                บ้านผาหมอน
221  เชียงใหม่                           บ้านแม่กลางหลวง
222  เชียงใหม่                                   บ้านถ้ำ
223  เชียงใหม่                                บ้านปางแดง
224  เชียงใหม่                                บ้านขอบด้ง
225  เชียงใหม่                                   บ้านยาง
226  เชียงใหม่                                บ้านกองแขก
227  เชียงใหม่                             บ้านแม่สาใหม่
228  เชียงใหม่                              บ้านแม่กำปอง
229  เชียงใหม่                                  เวียงแหง
230  เชียงใหม่                              บ้านต้นเกว๋น

## **นักท่องเที่ยวจะไปเที่ยวชุมชนเก่าในจังหวัดใดด้วย**

### K-itemset


In [28]:
transactions = data_mix.groupby('ชื่อย่านชุมชนเก่า')['รหัสจังหวัด'].apply(list)

In [29]:
transactions

ชื่อย่านชุมชนเก่า
กรือเซะ                  [th-pi]
กาดกองต้า                [th-lg]
การค้าถนนตราด-แหลมงอบ    [th-tt]
กูเบ                     [th-yl]
คลองน้อย                 [th-st]
                          ...   
โกตาบารู                 [th-yl]
โต๊ะบาหลิว               [th-kr]
ในไร่ คลองดาว            [th-kr]
ไม่มีข้อมูล              [th-bm]
ไหล่หิน                  [th-lg]
Name: รหัสจังหวัด, Length: 609, dtype: object

In [30]:
!pip install apyori

  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5974 sha256=cd2ad3bff5946215fe53ea122ce057951686e819c14005d42b35cadef3ee9b59
  Stored in directory: /root/.cache/pip/wheels/cb/f6/e1/57973c631d27efd1a2f375bd6a83b2a616c4021f24aab84080
Successfully built apyori


In [31]:
from apyori import apriori

In [32]:
frequentItemsets = list(apriori(transactions,min_support = 0.0015)) 
frequentItemsets

[RelationRecord(items=frozenset({'th-ac'}), support=0.0016420361247947454, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'th-ac'}), confidence=0.0016420361247947454, lift=1.0)]),
 RelationRecord(items=frozenset({'th-at'}), support=0.0049261083743842365, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'th-at'}), confidence=0.0049261083743842365, lift=1.0)]),
 RelationRecord(items=frozenset({'th-bk'}), support=0.006568144499178982, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'th-bk'}), confidence=0.006568144499178982, lift=1.0)]),
 RelationRecord(items=frozenset({'th-bm'}), support=0.0016420361247947454, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'th-bm'}), confidence=0.0016420361247947454, lift=1.0)]),
 RelationRecord(items=frozenset({'th-br'}), support=0.009852216748768473, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_a

In [33]:
frequentItemsets[-1]

RelationRecord(items=frozenset({'th-re', 'th-nr', 'th-br', 'th-ms'}), support=0.0016420361247947454, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'th-re', 'th-nr', 'th-br', 'th-ms'}), confidence=0.0016420361247947454, lift=1.0), OrderedStatistic(items_base=frozenset({'th-br'}), items_add=frozenset({'th-nr', 'th-re', 'th-ms'}), confidence=0.16666666666666666, lift=101.5), OrderedStatistic(items_base=frozenset({'th-ms'}), items_add=frozenset({'th-nr', 'th-re', 'th-br'}), confidence=0.11111111111111112, lift=67.66666666666667), OrderedStatistic(items_base=frozenset({'th-nr'}), items_add=frozenset({'th-re', 'th-br', 'th-ms'}), confidence=0.25, lift=152.25), OrderedStatistic(items_base=frozenset({'th-re'}), items_add=frozenset({'th-nr', 'th-br', 'th-ms'}), confidence=0.11111111111111112, lift=67.66666666666667), OrderedStatistic(items_base=frozenset({'th-br', 'th-ms'}), items_add=frozenset({'th-nr', 'th-re'}), confidence=1.0, lift=609.0), OrderedStatisti

RelationRecord(items=frozenset({'th-nr', 'th-br', 'th-ms', 'th-re'}), support=0.0016420361247947454, 

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'th-nr', 'th-br', 'th-ms', 'th-re'}), confidence=0.0016420361247947454, lift=1.0), 

OrderedStatistic(items_base=frozenset({'th-br'}), items_add=frozenset({'th-nr', 'th-ms', 'th-re'}), confidence=0.16666666666666666, lift=101.5), 

OrderedStatistic(items_base=frozenset({'th-ms'}), items_add=frozenset({'th-nr', 'th-br', 'th-re'}), confidence=0.11111111111111112, lift=67.66666666666667), 

OrderedStatistic(items_base=frozenset({'th-nr'}), items_add=frozenset({'th-br', 'th-ms', 'th-re'}), confidence=0.25, lift=152.25), 

OrderedStatistic(items_base=frozenset({'th-re'}), items_add=frozenset({'th-nr', 'th-br', 'th-ms'}), confidence=0.11111111111111112, lift=67.66666666666667), 

OrderedStatistic(items_base=frozenset({'th-br', 'th-ms'}), items_add=frozenset({'th-nr', 'th-re'}), confidence=1.0, lift=609.0), 

OrderedStatistic(items_base=frozenset({'th-nr', 'th-br'}), items_add=frozenset({'th-ms', 'th-re'}), confidence=1.0, lift=609.0), 

OrderedStatistic(items_base=frozenset({'th-br', 'th-re'}), items_add=frozenset({'th-nr', 'th-ms'}), confidence=1.0, lift=609.0), 

OrderedStatistic(items_base=frozenset({'th-nr', 'th-ms'}), items_add=frozenset({'th-br', 'th-re'}), confidence=1.0, lift=609.0), 

OrderedStatistic(items_base=frozenset({'th-ms', 'th-re'}), items_add=frozenset({'th-nr', 'th-br'}), confidence=1.0, lift=609.0), 

OrderedStatistic(items_base=frozenset({'th-nr', 'th-re'}), items_add=frozenset({'th-br', 'th-ms'}), confidence=1.0, lift=609.0), 

OrderedStatistic(items_base=frozenset({'th-nr', 'th-br', 'th-ms'}), items_add=frozenset({'th-re'}), confidence=1.0, lift=67.66666666666667), 

OrderedStatistic(items_base=frozenset({'th-br', 'th-ms', 'th-re'}), items_add=frozenset({'th-nr'}), confidence=1.0, lift=152.25), 

OrderedStatistic(items_base=frozenset({'th-nr', 'th-br', 'th-re'}), items_add=frozenset({'th-ms'}), confidence=1.0, lift=67.66666666666667), 

OrderedStatistic(items_base=frozenset({'th-nr', 'th-ms', 'th-re'}), items_add=frozenset({'th-br'}), confidence=1.0, lift=101.5)])

**OrderedStatistic(items_base=frozenset({'th-nr', 'th-ms'}), items_add=frozenset({'th-br', 'th-re'}), confidence=1.0, lift=609.0)**

In [34]:
data_mix[data_mix['รหัสจังหวัด']=='th-nr']

จังหวัด  ...                                        ลักษณะสำคัญ
348  นครราชสีมา  ...  ชุมชนเมืองเก่า เป็นย่านพาณิชยกรรมเก่า ประกอบด้...
349  นครราชสีมา  ...  ชุมชนเมืองเก่าพิมาย เป็นย่านพาณิชยกรรมเก่า ประ...
350  นครราชสีมา  ...  ชุมชนย่านตลาดปักธงชัย เป็นย่านพาณิชยกรรมเก่า ป...
351  นครราชสีมา  ...  ชุมชนย่านการค้าสีคิ้ว เป็นย่านพาณิชยกรรมเก่า ป...

[4 rows x 23 columns]

In [35]:
data_mix[data_mix['รหัสจังหวัด']=='th-ms']

จังหวัด  ...                                        ลักษณะสำคัญ
429  มหาสารคาม  ...  ชุมชนเมืองเก่ามหาสารคาม เป็นย่านพาณิชยกรรมเก่า...
430  มหาสารคาม  ...  ชุมชนเมืองโบราณบ้านเชียงเหียน เป็นชุมชนที่ตั้ง...
431  มหาสารคาม  ...  ชุมชนบ้านหม้อ เป็นชุมชนขนาดกลางตั้งอยู่บนสภาพภ...
432  มหาสารคาม  ...  ชุมชนบ้านปรางค์กู่ เป็นชุมชนขนาดกลางตั้งอยู่บน...
433  มหาสารคาม  ...  ชุมชนเมืองโบราณบ้านสระ เป็นชุมชนที่ตั้งอยู่บนส...
434  มหาสารคาม  ...  ชุมชนญ้อภูไท ท่าขอนยาง เป็นชุมชนที่ตั้งอยู่บนพ...
435  มหาสารคาม  ...  ชุมชนพยัคฆภูมิพิสัย เป็นย่านพาณิชยกรรมเก่า ประ...
436  มหาสารคาม  ...  ชุมชนเมืองเก่า เป็นย่านพาณิชยกรรมเก่าของอำเภอโ...
437  มหาสารคาม  ...  ชุมชนบ้านเมืองเตา เป็นชุมชนที่ตั้งอยู่บนสภาพภู...

[9 rows x 23 columns]

In [36]:
data_mix[data_mix['รหัสจังหวัด']=='th-br']

จังหวัด  ...                                        ลักษณะสำคัญ
352  บุรีรัมย์  ...  ชุมชนคูเมืองเก่า เป็นย่านพาณิชยกรรมและที่พักอา...
353  บุรีรัมย์  ...  ชุมชนริมสถานีรถไฟ เป็นย่านพาณิชยกรรมเก่า ภายใน...
354  บุรีรัมย์  ...  ชุมชนประชาสันติสุข เป็นย่านพาณิชยกรรมเก่าในยุค...
355  บุรีรัมย์  ...  ชุมชนเมืองเก่า เป็นย่านพาณิชยกรรมและพักอาศัยเก...
356  บุรีรัมย์  ...  ชุมชนโดยรอบปราสาทหินเมืองต่ำ เป็นชุมชนพักอาศัย...
357  บุรีรัมย์  ...  ชุมชนพุทไธสง เป็นย่านพาณิชยกรรมเก่า ตั้งอยู่บน...

[6 rows x 23 columns]

In [37]:
data_mix[data_mix['รหัสจังหวัด']=='th-re']

จังหวัด  ...                                        ลักษณะสำคัญ
438  ร้อยเอ็ด  ...  ชุมชนเมืองเก่า เป็นชุมชนที่อยู่รายรอบคูเมืองโบ...
439  ร้อยเอ็ด  ...  ชุมชนสีแก้ว เป็นย่านชุมชนเก่าที่มีความทรงจำ/เห...
440  ร้อยเอ็ด  ...  ชุมชนเมืองบัว เป็นชุมชนโบราณที่มีคูน้ำคันดินล้...
441  ร้อยเอ็ด  ...  ชุมชนบ้านเมืองไพร เป็นย่านชุมชนเก่าที่มีความทร...
442  ร้อยเอ็ด  ...  ชุมชนเมืองเก่า เป็นย่านพาณิชยกรรมเก่า ประกอบด้...
443  ร้อยเอ็ด  ...  ชุมชนท่าม่วง เป็นชุมชนที่ตั้งอยู่ริมลำน้ำชี ตั...
444  ร้อยเอ็ด  ...  ชุมชนบ้านสังข์ เป็นชุมชนที่มีแหล่งน้ำธรรมชาติ ...
445  ร้อยเอ็ด  ...  ชุมชนบ้านน้ำคำ เป็นชุมชนที่ตั้งอยู่โดยรอบของโบ...
446  ร้อยเอ็ด  ...  ชุมชนบ้านสังข์ เป็นชุมชนที่มีแหล่งน้ำธรรมชาติห...
447  ร้อยเอ็ด  ...  ชุมชนธวัชดินแดง เป็นชุมชนที่มีแหล่งน้ำธรรมชาติ...
448  ร้อยเอ็ด  ...  ชุมชนฟืม (บ้านเขือง) เป็นชุมชนที่มีแหล่งน้ำธรร...

[11 rows x 23 columns]

**สรุป** ถ้าหากนักท่องเที่ยวไปท่องเที่ยวชุมชนเก่าในจังหวัด นครราชสีมา และ มหาสารคาม นักท่องเที่ยวจะไปท่องเที่ยวชุมชนเก่าในจังหวัด บุรีรัมย์ และ ร้อยเอ็ด ด้วย โดยมีค่า confidence = 100% ที่ค่า support = 0.1% ซึ่งมีค่า lift มากกว่า 1
  

### Decision tree

In [38]:
from sklearn import datasets

In [39]:
data_mix.to_csv('data_mix.csv')

In [68]:
col_names = ['จังหวัด','รหัสจังหวัด','2555visitor','2556visitor','2557visitor','2558visitor','2559visitor','2560visitor','2561visitor','2555income','2556income','2557income','2558income','2559income','2560income','2561income','ชื่อย่านชุมชนเก่า','ตำบล','อำเภอ','ภาค','East','North','ลักษณะสำคัญ']
data_mytree = pd.read_csv("data_mix.csv",header=None, names=col_names)

In [71]:
data_mytree

จังหวัด  ...                                        ลักษณะสำคัญ
NaN          จังหวัด  ...                                        ลักษณะสำคัญ
0.0    กรุงเทพมหานคร  ...                                        ไม่มีข้อมูล
1.0      สมุทรปราการ  ...  ชุมชนเมืองนครเขื่อนขันธ์ เป็นชุมชนย่านการค้าเก...
2.0      สมุทรปราการ  ...  ชุมชนตลาดบางพลี เป็นชุมชนย่านตลาดการค้าเก่าแก่...
3.0      สมุทรปราการ  ...  ชุมชนตลาดคลองสวน เป็นชุมชนย่านตลาดการค้า บริเว...
...              ...  ...                                                ...
609.0       นราธิวาส  ...  ดุซงญอ เป็นชุมชนตลาด และชุมชนความทรงจำ "ดุซงญอ...
610.0       นราธิวาส  ...  ตะโละมาเนาะ เป็นชุมชนเกษตรกรรม และชุมชนโดยรอบโ...
611.0       นราธิวาส  ...  ละหาร เป็นชุมชนเกษตรกรรม บ้านละหารตั้งอยู่ในอำ...
612.0       นราธิวาส  ...  ตลาดยี่งอ เป็นชุมชนตลาด "เมืองยี่งอ" อาจจะเรีย...
613.0       นราธิวาส  ...  ตลาดสุไหงปาดี เป็นชุมชนตลาด และชุมชนโดยรอบทางร...

[615 rows x 23 columns]

In [73]:
feature_cols = ['จังหวัด','รหัสจังหวัด','2555visitor','2556visitor','2557visitor','2558visitor','2559visitor','2560visitor','2561visitor','2555income','2556income','2557income','2558income','2559income','2560income','2561income','ชื่อย่านชุมชนเก่า','ตำบล','อำเภอ','ภาค','East','North','ลักษณะสำคัญ']
X = data_mytree[feature_cols] # Features
y = data_mytree.รหัสจังหวัด # Target variable

## Train Model

###import (เรียกใช้ algoritihm ที่เราต้องการ)

In [57]:
from sklearn.tree import DecisionTreeClassifier

###define (การกำหนด parameters ให้กับ model)

In [58]:
mytree = DecisionTreeClassifier()

###train (ฝึกสอนตัวแบบ)

In [59]:
mytree.fit(X,Y)

ValueError: ignored